src/Streamly/Internal/Prelude.hs

```diff
@@ -783,12 +782,9 @@ iterate step = K.fromStream . go
 -- /Since: 0.7.0 (signature change)/
 --
 -- /Since: 0.1.2/
+{-# INLINE_EARLY iterateM #-}
 iterateM :: (IsStream t, MonadAsync m) => (a -> m a) -> m a -> t m a
-iterateM step = go
-    where
-    go s = K.mkStream $ \st stp sng yld -> do
-        next <- s
-        K.foldStreamShared st stp sng yld (return next |: go (step next))
+iterateM step = fromStreamS . S.iterateM step
```

Comment:
This is not concurrent any more. It will pick up only StreamD implementation which is not concurrent. We need to use the StreamK implementation and use a rewrite rule for the serial case, see repeatM implementation above.

Reply:
We should perhaps have a test to check for concurrency so that we do not miss something like this in future.